leptos/src/for_loop.rs

```diff
@@ -76,6 +81,83 @@ where
     move || keyed(each(), key.clone(), children.clone())
 }
 
+/// Iterates over children and displays them, keyed by the `key` function given.
+///
+/// Compared with For, it has an additional index parameter, which can be used to obtain the current index in real time.
+///
+/// This is much more efficient than naively iterating over nodes with `.iter().map(|n| view! { ... })...`,
+/// as it avoids re-creating DOM nodes that are not being changed.
+///
+/// ```
+/// # use leptos::*;
+///
+/// #[derive(Copy, Clone, Debug, PartialEq, Eq)]
+/// struct Counter {
+///   id: usize,
+///   count: RwSignal<i32>
+/// }
+///
+/// #[component]
+/// fn Counters() -> impl IntoView {
+///   let (counters, set_counters) = create_signal::<Vec<Counter>>(vec![]);
+///
+///   view! {
+///     <div>
+///       <For
+///         // a function that returns the items we're iterating over; a signal is fine
+///         each=move || counters.get()
+///         // a unique key for each item
+///         key=|counter| counter.id
+///         // renders each item to a view
+///         children=move |index: MaybeSignal<usize>, counter: Counter| {
+///           view! {
+///             <button>{move || index.get()} ". Value: " {move || counter.count.get()}</button>
+///           }
+///         }
+///       />
+///     </div>
+///   }
+/// }
+/// ```
+#[cfg_attr(feature = "tracing", tracing::instrument(level = "trace", skip_all))]
+#[component]
+pub fn ForEnumerate<IF, I, T, EF, N, KF, K>(
+    /// Items over which the component should iterate.
+    each: IF,
+    /// A key function that will be applied to each item.
+    key: KF,
+    /// A function that takes the index and the item, and returns the view that will be displayed for each item.
+    children: EF,
+) -> impl IntoView
+where
+    IF: Fn() -> I + Send + 'static,
+    I: IntoIterator<Item = T> + Send + 'static,
+    EF: Fn(MaybeSignal<usize>, T) -> N + Send + Clone + 'static,
```

Comment:
Why does it take `MaybeSignal`, when we know that it is always a signal? Why not just have it take a `ReadSignal`?